compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs

```diff
@@ -907,11 +907,20 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             };
             let ty = self.normalize(expr.span, ty);
             if self.can_coerce(found, ty) {
+                let mut node_id = self.tcx.hir().parent_id(expr.hir_id);
+                let mut parent_id_iter = self.tcx.hir().parent_iter(node_id);
+                while let Some((id, _)) = parent_id_iter.next() && self.tcx.hir().span(id).from_expansion() {
+                    node_id = id;
+                }
+                let parent = self.tcx.hir().get_enclosing_scope(node_id).unwrap_or(node_id);
+                let parent_span = self.tcx.hir().span(parent);
+                let span = expr.span.find_ancestor_inside(parent_span).unwrap_or(expr.span);
```

Comment:
I wonder if we should have a `Span::find_suggestable_ancestor` which walks the expansion tree up, until we are not in an external macro any more. With the definition of external macros given by `rustc_middle::lint::in_external_macro`.

At least, the two instances of this code deserve a helper function somewhere.

Reply:
> At least, the two instances of this code deserve a helper function somewhere.

I extracted this code into a helper function `find_suggestable_ancestor_span` inside `rustc_hir_typeck/src/fn_ctxt/suggestions.rs` to avoid the duplication.